### DISCUSSION TITLE
Number of bytes required to encode `F32`

### DISCUSSION THREAD

**Discussion by ThibFrgsGmz** [esa/asn1scc#243] [Q&A] [answered]
Hi,

I am structuring my messages in `ASN.1` and I see an increase of about 50% in the structures where there are `F32s`.

If I define a message containing only one `F32`, I see that it takes `13 bytes` to encode it.

Is this normal? I use the default encoding `uPER`.

What is the procedure to get only `32 bits` to encode the floats please?

**Comment by maxime-esa** [comment, marked as answer]
Assuming your type is defined with this range (corresponding to a 32bits float):
``` 
MyF32 ::= REAL (-3.4e38 .. 3.4e38)
```

For the binary encoding you can choose between various schemes when you invoke the compiler: either uPER (ASN.1's unaligned Packed Encoding Rules) with the `-uPER` flag or ACN with `-ACN`.

For real numbers, PER is not an optimal choice because it needs at least 13 bytes whatever precision you need. In fact it theoretically provides "infinite precision" which your do not need most of the time.

So to solve your issue, first create an ACN file for your grammar (`f32.asn`) like this:
```
$ asn1scc -ACND f32.asn
```

Edit the file that was generated (`f32.acn`) and add the following directive to specify that you want a single-precision encoding using 32 bits only:

```
MyF32[encoding IEEE754-1985-32]
```

... and check the resulting encoding layout:
```
$ asn1scc -icdAcn out.html -ACN f32.asn f32.acn
```
The resulting file (`out.html`) should contain this nice table:

![image](https://user-images.githubusercontent.com/1416005/191588884-f093d2a6-c375-44d6-bce2-302b817715e6.png)

In order to generate automatic encoders and decoders for this scheme in C, use the `-c` flag. 

The complete user manual detailing all ACN directives is here:
https://taste.tuxfamily.org/wiki/index.php?title=Technical_topic:_ASN.1_-_An_introduction_to_ACN

**Reply by usr3-1415** [reply]
Moreover, you can specify the order of bytes in the encoding stream 

MyF32[encoding IEEE754-1985-32, endianness little]
or
MyF32[encoding IEEE754-1985-32, endianness big]

**Reply by ThibFrgsGmz** [reply]
Thanks guys!
